Pytest documentation + build tweaks

QuLogic opened the pull request:
I've added the developer documentation for setting up pytest and removed a few more cases of nose in the setup, package metadata, etc.

For the tweaks:
* Fix `tests.py` when run with a module path instead of a file path.
* Close figures before running `image_comparison` because that decorator expects all figures when it's done to correspond to figures from the test, but if never checks for/ignores any already open figures.
* Delete `test_mouseclicks.py` example because it's getting collected and run by pytest and since it's not really written as a test it causes an extra figure to be opened which breaks any image tests if run directly (modulo the previous fix.) Also copy the use of `event.dblclick` into the documentation so it's not gone completely.
* Use `importorskip` everywhere so that the skip reasons are always the same, since it's used in some places and not others.

Fixes #8015.

QuLogic commented:
Added one more tweak to use `importorskip` everywhere so that the skip reasons are always the same, since it's used in some places and not others.

QuLogic commented:
I've added the developer documentation for setting up pytest and removed a few more cases of nose in the setup, package metadata, etc.

anntzer commented:
Could we document the pytest interface rather than `tests.py`?  I think it basically comes down to documenting `-m "not network"` instead of `--no-network`, the rest is the same.

After a quick look at `tests.py`, it also seems that there are some customizations (warnings filters, recursion limit) which should perhaps be moved to `conftest.py`?

dstansby commented on tox.ini:
```diff
@@ -12,7 +12,6 @@ commands =
     sh -c 'rm -f $HOME/.matplotlib/fontList*'
```
Add python version 3.6 to `envlist` above

QuLogic replied:
Added.

dstansby commented on lib/matplotlib/tests/test_axes.py:
```diff
@@ -4673,10 +4667,7 @@ def test_pandas_errorbar_indexing():
-        import pandas as pd
-    except ImportError:
-        pytest.skip("Pandas not installed")
+    pd = pytest.importorskip('pandas')
```
Is there a minimum required version of pandas needed for the pandas tests to work?

phobson replied:
It doesn't appear that there is. Creating a `Series` from a list and `DataFrame` a from a dictionary of lists/arrays should work with versions of pandas several years old at this point.

dstansby commented on examples/event_handling/test_mouseclicks.py:
```diff
@@ -1,32 +0,0 @@
-from __future__ import print_function
```
Also need to get rid of `exammples/event_handling/test_mouseclicks.rst`

jenshnielsen replied:
Isn't that file autogenerated from the example? It's not in version control as far as I can see

dstansby replied:
:man_facepalming: so it is

dstansby commented on doc/devel/testing.rst:
```diff
@@ -53,58 +56,72 @@ value.
 .. hint::
 
-   To run the tests you need to install nose and mock if using python 2.7::
+   To run the tests you need to install pytest and mock if using python 2.7::
```
Everything in this hint is covered above, so I think it can just be deleted

dstansby commented on doc/devel/testing.rst:
```diff
@@ -53,58 +56,72 @@ value.
-documentation for supported arguments. Some of the more important ones are given
-here:
+Additional arguments are passed on to pytest. See the pytest documentation for
+supported arguments. Some of the more important ones are given here:
```
Might be good to put a link to the pytest documents that list supported arguments

dstansby commented on doc/devel/testing.rst:
```diff
@@ -53,58 +56,72 @@ value.
+colons, e.g., (this is assuming the test is installed)::
 
-  python tests.py matplotlib.tests.test_simplification:test_clipping
+  python tests.py matplotlib.tests.test_simplification::test_clipping
```
Is there any point documenting the method where tests have to be installed, if the method below (which doesn't require installed tests) works?

QuLogic replied:
I don't know; does anyone use this method?

On further though, using non-installed tests might be broken when trying `mpl_toolkits` tests because they import stuff from `matplotlib.tests`. I can move things around to get that working again.

QuLogic commented:
I updated the documentation to follow the requests here and also moved some bits around so that tests work without being installed. Everything but the determinism test in the SVG backend works (because that one imports itself, which doesn't exist when not installed); not too sure how to fix that one, but it's not a high priority since not-installed tests never worked before.

NelleV commented on doc/devel/contributing.rst:
```diff
@@ -120,14 +120,16 @@ environment is set up properly::
-  <https://docs.python.org/dev/library/unittest.mock.html>`_ (if python < 3.3), `Ghostscript
-  <https://www.ghostscript.com/>`_, `Inkscape <https://inkscape.org>`_
+  **Additional dependencies for testing**: pytest_ (version 3.0 or later),
+  mock_ (if python < 3.3), Ghostscript_, Inkscape_
```
:+1:

dstansby replied:
Python 3 < 3.3 isn't supported but matplotlib, so might make more sense to change this to "if python 2.7"

NelleV commented on doc/devel/testing.rst:
```diff
@@ -4,31 +4,34 @@
-please ignore it while we consolidate our testing to these locations.)
-
-.. _nose: https://nose.readthedocs.io/en/latest/
+Matplotlib has a testing infrastructure based on pytest_, making it easy to
```
Here are suggestions:
```
Matplotlib's testing infrastucture relies on pytest_. Test files are place in [path to the folder], and testing utility functions are in :mod:`matplotlib.testing`.
```
I'd drop the "making it easy to write new tests", as this is a judgement call that I also totally disagree with.
I'd also remove the "There is other old testing cruft around…"

QuLogic replied:
Changed; also switched to the file paths instead of modules since we're talking about developers who probably want to edit the existing files.

NelleV commented on doc/devel/testing.rst:
```diff
@@ -53,58 +56,66 @@ value.
-Running the tests is simple. Make sure you have nose installed and run::
+Running the tests is simple. Make sure you have pytest installed and run::
+
+   py.test
```
:+1:

dstansby replied:
See above comment about `py.test` vs. `pytest`

NelleV commented on doc/devel/testing.rst:
```diff
@@ -53,58 +56,66 @@ value.
+
+or::
 
    python tests.py
```
@tacaswell can we remove this from our documentation? :(

anntzer replied:
I agree, attn @tacaswell.

NelleV commented on doc/devel/testing.rst:
```diff
@@ -4,31 +4,34 @@
-    information
-  - `pep8 <http://pep8.readthedocs.io/en/latest>`_ to test coding standards
+  - pytest-cov_ to collect coverage information
+  - pytest-pep8_ to test coding standards
```
I would add all the optional dependencies here:
   -  pytest-xdist
   -  pytest-timeout_

QuLogic replied:
Added both.

NelleV commented on doc/devel/testing.rst:
```diff
@@ -53,58 +56,66 @@ value.
+not need to be installed, but Matplotlib should be)::
 
-  python tests.py matplotlib.tests.test_simplification:test_clipping
+  py.test lib/matplotlib/tests/test_simplification.py::test_clipping
```
:+1:

NelleV commented on doc/devel/testing.rst:
```diff
@@ -53,58 +56,66 @@ value.
-  python tests.py matplotlib.tests.test_simplification:test_clipping
+  py.test lib/matplotlib/tests/test_simplification.py::test_clipping
+
+or, if tests are installed, a dot-separated path to the module, optionally
```
I would remove this section. We only need to document one way to do this.

NelleV commented on doc/devel/testing.rst:
```diff
@@ -53,58 +56,66 @@ value.
+function :func:`matplotlib.test`::
 
   import matplotlib
   matplotlib.test()
```
I'd remove this section from our documentation. I think we should document (and have) only one way to run the tests.

attn @tacaswell (as I believe Thomas will disagree with me on that one).

NelleV commented on doc/devel/testing.rst:
```diff
@@ -113,30 +124,20 @@ Writing a simple test
-      """
-      fig = figure()
-      ...
+Tests that have side effects that need to be cleaned up, such as created
```
That paragraph is very unclear to me. Here is a suggestions.

"Some tests have side effects that need to be cleaned up after their execution (such as created figures or modifiec rc params). The pytest fixture XXX will automatically clean these up:

**add short example showing how to use the fixture**"

QuLogic replied:
There's no need for an example because the fixture gets used automatically; I've swapped the text around as suggested.

NelleV commented on doc/users/event_handling.rst:
```diff
@@ -29,13 +29,13 @@ connect your function to the event manager, which is part of the
 
-    fig = plt.figure()
-    ax = fig.add_subplot(111)
+    fig, ax = plt.subplots()
```
:+1:

madphysicist commented:
Thanks for doing this. This is very helpful. Both the docs and switching to pytest. There is a bit of a learning curve with it, but it seems to be much more flexible than nose.

QuLogic commented on INSTALL:
```diff
@@ -106,12 +106,10 @@ or example code.
 distribution, download the :file:`*.tar.gz` file and look in the
 :file:`examples` subdirectory.
-To run the test suite, copy the :file:`lib\\matplotlib\\tests` and
+To run the test suite, extract the :file:`lib\\matplotlib\\tests` or
```
I split things into bullet points; hopefully that's clearer.

QuLogic commented on README.rst:
```diff
@@ -28,9 +28,9 @@ Or from the Python interpreter::
-more information. Note that the test suite requires nose and on Python 2.7 mock
-which are not installed by default. Please install with pip or your package
-manager of choice.
+more information. Note that the test suite requires pytest and on Python 2.7
```
Changed.

QuLogic commented:
I've removed an extra 3.5 build on Travis because it's pretty redundant; hopefully that will get us to finished CI just a little bit faster.

NelleV commented:
LGTM :+1:

anntzer commented on doc/devel/testing.rst:
```diff
@@ -53,58 +58,64 @@ value.
-An alternative implementation that does not look at command line
-arguments works from within Python is to run the tests from the
-matplotlib library function :func:`matplotlib.test`::
+  PYTHONHASHSEED=0 py.test --verbose -n 5
```
Can we just specify a minimum version of pytest-xdist where this is no longer an issue?  It looks like it is related to https://bitbucket.org/pytest-dev/pytest/issues/346/pytest-xdist-and-python-33-is-sort-of, which has been resolved in 2013.

QuLogic replied:
Unfortunately, it's not fixed yet: pytest-dev/pytest#920

I just PR'd the same fix to the pandas build, so I know that recent versions still don't work.

dstansby commented on INSTALL:
```diff
@@ -106,14 +106,15 @@ or example code.
+   `mock <https://pypi.python.org/pypi/mock>`_, Pillow, MiKTeX, GhostScript,
+   ffmpeg, avconv, mencoder, ImageMagick, and `Inkscape
+   <https://inkscape.org/>`_;
+ * run ``py.test path\\to\\tests\\directory``.
```
- According to https://github.com/pytest-dev/pytest/issues/1629#issue-161422224 , `pytest` is preferred over `py.test` as of 3.0
- I don't think you need the extra `path\\to\\tests\\directory` bit; if you do then line 23 of `README.rst` should be made consistent

QuLogic replied:
This is about separate uninstalled tests; line 23 in `README.rst` is about running from the full source directory.

dstansby commented on README.rst:
```diff
@@ -20,17 +20,16 @@ Testing
 After installation, you can launch the test suite::
 
-  python tests.py
+  py.test
```
See previous comment about `py.test` vs. `pytest`

dstansby commented on doc/devel/testing.rst:
```diff
@@ -53,58 +58,64 @@ value.
+
+  py.test --pyargs matplotlib.tests.test_simplification::test_clipping
 
 If you want to run the full test suite, but want to save wall time try
```
What is "wall time"?

QuLogic replied:
Time on a clock on the wall, as opposed to cpu time.

dstansby commented:
Since all the lines I commented on are technically true and easily changed later, I'm going to merge (finally!), thanks for all the work on this @QuLogic 👍
